### PR TITLE
blueprint with two/more tiers

### PR DESCRIPTION
**Pull request by jesuspg**
#### Reviewers
@hmunfru @jframos 

#### Description
https://jira.tid.es/browse/CLAUDIA-4678

#### Test
Tested locally with experimentacion testbed

### CODE REVIEW

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/dao/EnvironmentInstanceDao.java [hunk 1]
```diff
@@ -22,28 +22,26 @@
...
+    /**
+     * Find the applicationInstancs that matches with the given criteria.
+     * 
+     * @param criteria
+     *            the search criteria
+     * @return the list of elements that match with the criteria.
+     */
+    List<EnvironmentInstance> findByCriteria(EnvironmentInstanceSearchCriteria criteria);
+
+    EnvironmentInstance load(String name, String vdc) throws EntityNotFoundException;
```
Comment (hmunfru) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/dao/EnvironmentInstanceDao.java]: ya no hace falta el load for delete?

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/dao/impl/EnvironmentInstanceDaoJpaImpl.java [hunk 2]
```diff
@@ -110,14 +103,13 @@ private EnvironmentInstance findByEnvironmentInstanceName(String envInstanceName
             throws EntityNotFoundException {
 
         Query query = getEntityManager().createQuery(
-                "select p from EnvironmentInstance"
-                        + " p join fetch p.tierInstances where p.blueprintName = :blueprintName and p.vdc =:vdc");
+                "select p from EnvironmentInstance" + " p where p.blueprintName = :blueprintName and p.vdc =:vdc");
```
Comment (hmunfru) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/dao/impl/EnvironmentInstanceDaoJpaImpl.java]: lo mismo no te hace falta ese + no?

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/dao/impl/EnvironmentInstanceDaoJpaImpl.java [hunk 3]
```diff
@@ -110,14 +103,13 @@ private EnvironmentInstance findByEnvironmentInstanceName(String envInstanceName
...
-                "select p from EnvironmentInstance"
-                        + " p join fetch p.tierInstances where p.blueprintName = :blueprintName and p.vdc =:vdc");
+                "select p from EnvironmentInstance" + " p where p.blueprintName = :blueprintName and p.vdc =:vdc");
         query.setParameter("blueprintName", envInstanceName);
         query.setParameter("vdc", vdc);
         EnvironmentInstance environmentInstance = null;
         try {
             environmentInstance = (EnvironmentInstance) query.getResultList().get(0);
-            getEntityManager().flush();
+            environmentInstance.getTierInstances().size();
```
Comment (hmunfru) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/dao/impl/EnvironmentInstanceDaoJpaImpl.java]: por que pones el size(), para que lo cargue?

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java [hunk 4]
```diff
@@ -1,675 +1,695 @@
...
+            log.info("error to find environment instance " + e.getMessage());
+            throw new EntityNotFoundException(EnvironmentInstance.class, "vdc", vdc);
+        }
+        if (!instance.getVdc().equals(vdc)) {
+            throw new EntityNotFoundException(EnvironmentInstance.class, "vdc", vdc);
+        }
+        return instance;
+    }
+
+    public EnvironmentInstance loadWithTiers(String vdc, String name) throws EntityNotFoundException {
```
Comment (hmunfru) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java]: que diferencia hay entre load y loadwithtiers?

File: rest-api/src/main/java/com/telefonica/euro_iaas/paasmanager/rest/resources/EnvironmentInstanceResourceImpl.java [hunk 5]
```diff
@@ -231,7 +231,7 @@ public Task destroy(String org, String vdc, String name, String callback) throws
         log.info("Destroy env isntna " + name + " vdc " + vdc);
         EnvironmentInstance environmentInstance = null;
         try {
-            environmentInstance = environmentInstanceManager.load(vdc, name);
+            environmentInstance = environmentInstanceManager.loadWithTiers(vdc, name);
```
Comment (hmunfru) [inline rest-api/src/main/java/com/telefonica/euro_iaas/paasmanager/rest/resources/EnvironmentInstanceResourceImpl.java]: el codigo parece el mismo el load que el loadiwthtiers

**Comment by hmunfru** [comment]
LGTM

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/dao/impl/EnvironmentInstanceDaoJpaImpl.java (hunk 3 above)
Comment (jesuspg) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/dao/impl/EnvironmentInstanceDaoJpaImpl.java]: si

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/dao/EnvironmentInstanceDao.java (hunk 1 above)
Comment (jesuspg) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/dao/EnvironmentInstanceDao.java]: loadForDelete llamaba a findByEnvironmentInstanceName, osea que loadForDelete era igual al load

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/dao/impl/TierDaoJpaImpl.java [hunk 6]
```diff
@@ -226,8 +226,8 @@ public Tier loadComplete(Tier newTier) throws EntityNotFoundException {
         Tier tier = null;
         try {
             tier = (Tier) query.getResultList().get(0);
-            tier.getNetworks();
-            tier.getProductReleases();
+            tier.getNetworks().size();
+            tier.getProductReleases().size();
             tier.getSecurityGroup();
         } catch (Exception e) {
             String message = " Tier don't exist in database ";
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/dao/impl/TierDaoJpaImpl.java]: Tier does not exist in database

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java [hunk 7]
```diff
@@ -1,675 +1,678 @@
...
+            environmentInstance.setStatus(Status.ERROR);
+            environmentInstanceDao.update(environmentInstance);
+            throw new ProductInstallatorException(e);
+        } catch (NotUniqueResultException e) {
+            environmentInstance.setStatus(Status.ERROR);
+            environmentInstanceDao.update(environmentInstance);
+            throw new ProductInstallatorException(e);
+        }
+
+        log.info("Is the environmetn federated ? ");
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java]: Is the environment federated ?

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java [hunk 8]
```diff
@@ -1,675 +1,678 @@
...
+
+        environmentInstance.setStatus(Status.INSTALLED);
+        environmentInstanceDao.update(environmentInstance);
+
+        infrastructureManager.StartStopScalability(claudiaData, bScalableEnvironment);
+
+        environmentInstance.setStatus(Status.INSTALLED);
+        environmentInstanceDao.update(environmentInstance);
+
+        log.info("Enviroment Instance installed correctly");
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java]: Environment Instance installed correctly

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java [hunk 9]
```diff
@@ -1,675 +1,678 @@
...
+                                + pie.getMessage();
+                        tierInstance.setStatus(Status.ERROR);
+                        tierInstanceDao.update(tierInstance);
+                        log.error(message);
+                        throw new ProductInstallatorException(message, pie);
+                    }
+                }
+
+                if (state && tierInstance.getNumberReplica() == 1) {
+                    log.info("Setup scalabiliy ");
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java]: Setup scalability

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java [hunk 10]
```diff
@@ -1,675 +1,678 @@
...
+
+    /**
+     * It destroy the environment.
+     * 
+     * @param claudiaData
+     * @param envInstance
+     * @throws Exception
+     */
+    public void destroy(ClaudiaData claudiaData, EnvironmentInstance envInstance) throws Exception {
+        log.info("Destroying enviornment isntance " + envInstance.getBlueprintName() + " with environment "
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java]: Destroying environment instance

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java [hunk 11]
```diff
@@ -1,675 +1,678 @@
...
+        try {
+            envInstance.setStatus(Status.UNINSTALLING);
+            envInstance = environmentInstanceDao.update(envInstance);
+            for (int i = 0; i < envInstance.getTierInstances().size(); i++) {
+                deleteVM(claudiaData, envInstance);
+            }
+
+        } catch (NullPointerException ne) {
+            log.info("Environment Instance " + envInstance.getBlueprintName()
+                    + " does not have any TierInstances associated");
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java]: does not have any TierInstance associated

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java [hunk 12]
```diff
@@ -1,675 +1,678 @@
...
+        }
+        tierInstanceManager.remove(tierInstance);
+    }
+
+    // PRVATE METHODS
+
+    private void deleteVM(ClaudiaData claudiaData, EnvironmentInstance envInstance) throws InvalidEntityException {
+        // Borrado de VMs
+        try {
+            log.info("Deleting Virtual Machines for environmetn instance " + envInstance.getBlueprintName());
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java]: Deleting Virtual Machines for environment instance

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java [hunk 13]
```diff
@@ -1,675 +1,678 @@
...
+                    tierDB.setProductReleases(pReleases);
+                    tiers.add(tierDB);
+                }
+                environment.setTiers(null);
+                environment.setTiers(tiers);
+
+                return environment;
+            } catch (Exception e1) {
+                e1.printStackTrace();
+                log.warn("Error to load env " + e1.getMessage());
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java]: Error loading env

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java [hunk 14]
```diff
@@ -1,675 +1,678 @@
...
+                environment.setOvf(env.getOvf());
+                environment = environmentManager.update(environment);
+            }
+            return environment;
+        } catch (EntityNotFoundException e1) {
+            try {
+                environment = environmentManager.create(claudiaData, env);
+            } catch (InvalidEnvironmentRequestException e) {
+                // TODO Auto-generated catch block
+                String errorMessage = " Error to create the environment . " + environment.getName() + ". " + "Desc: "
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java]: Error creating the environment

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java [hunk 15]
```diff
@@ -1,675 +1,678 @@
...
+                        + e.getMessage();
+                log.error(errorMessage);
+                throw new InvalidEntityException(Environment.class, e);
+            } catch (Exception e) {
+                // TODO Auto-generated catch block
+                String errorMessage = " Error to create the environment . " + environment.getName() + ". " + "Desc: "
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java]: idem

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/dao/impl/EnvironmentDaoJpaImpl.java [hunk 16]
```diff
@@ -61,15 +62,16 @@ public Environment load(String envName) throws EntityNotFoundException {
...
-                "select p from Environment p left join " + "fetch p.tiers where p.name = :name and p.vdc = :vdc");
+                "select p from Environment p left join fetch p.tiers where p.name = :name and p.vdc = :vdc");
+
         query.setParameter("name", envName);
         query.setParameter("vdc", vdc);
         Environment environment = null;
         try {
             environment = (Environment) query.getResultList().get(0);
-            log.info("" + Hibernate.isInitialized(environment));
-            Hibernate.initialize(environment);
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/dao/impl/EnvironmentDaoJpaImpl.java]: You do not need to initialise the variable with Hibernate at all?

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java [hunk 17]
```diff
@@ -1,675 +1,678 @@
...
+        environmentInstance.setStatus(Status.INIT);
+
+        environmentInstance = insertEnvironmentInstanceInDatabase(environmentInstance);
+
+        environmentManager.loadNetworks(environment);
+        if (environment.isNetworkFederated()) {
+            try {
+                updateFederatedNetworks(claudiaData, environment);
+            } catch (Exception e) {
+                log.warn("It is not possible to update the federates networks");
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java]: It is not possible to update the federated networks

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java [hunk 18]
```diff
@@ -1,675 +1,678 @@
...
+                throw new InfrastructureException(e);
+            }
+        }
+
+        environmentInstance.setStatus(Status.INSTALLED);
+        environmentInstanceDao.update(environmentInstance);
+
+        infrastructureManager.StartStopScalability(claudiaData, bScalableEnvironment);
+
+        environmentInstance.setStatus(Status.INSTALLED);
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java]: Why do you assign twice the status.installed

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java [hunk 19]
```diff
@@ -1,675 +1,678 @@
...
+                if (state && tierInstance.getNumberReplica() == 1) {
+                    log.info("Setup scalabiliy ");
+                    String image_Name = infrastructureManager.ImageScalability(claudiaData, tierInstance);
+                }
+
+                if (state && tierInstance.getNumberReplica() > 1) {
+                    log.info("Updating OVF replica more than 1 ");
+                    if (!newOVF.equals(" "))
+                        tierInstance.setOvf(newOVF);
+                }
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java]: Why do you not change this code block by:


if (state) {
    if ( tierInstance.getNumberReplica() == 1) {
        log.info("Setup scalabiliy ");
       String image_Name = infrastructureManager.ImageScalability(claudiaData, tierInstance);
    }

    if (tierInstance.getNumberReplica() > 1) {
        log.info("Updating OVF replica more than 1 ");

        if (!newOVF.equals(" "))  {
            tierInstance.setOvf(newOVF);
        }
    }
}

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java [hunk 20]
```diff
@@ -1,675 +1,678 @@
...
+
+                if (state && tierInstance.getNumberReplica() == 1) {
+                    log.info("Setup scalabiliy ");
+                    String image_Name = infrastructureManager.ImageScalability(claudiaData, tierInstance);
+                }
+
+                if (state && tierInstance.getNumberReplica() > 1) {
+                    log.info("Updating OVF replica more than 1 ");
+                    if (!newOVF.equals(" "))
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java]: Do not forget that you need to add {, } to all if sentences...

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java [hunk 21]
```diff
@@ -1,675 +1,678 @@
...
+        List<TierInstance> tierInstancesSDC = envInstance.getTierInstances();
+        envInstance.setTierInstances(null);
+
+        for (int i = 0; i < tierInstancesSDC.size(); i++) {
+            // delete data on SDC
+            TierInstance tierInstance = tierInstancesSDC.get(i);
+            try {
+                deleteTierOnSDC(claudiaData, tierInstance, error);
+            } catch (Exception e) {
+                String errorMsg = "Error deleting node from Node Manager : " + tierInstance.getVM().getFqn() + ""
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java]: adding only "" is adding nothing...

File: model/src/main/java/com/telefonica/euro_iaas/paasmanager/model/Tier.java [hunk 22]
```diff
@@ -82,16 +84,11 @@
     private String payload;
 
     @ManyToMany(fetch = FetchType.LAZY)
-    @JoinTable(name = "tier_has_productReleases", joinColumns =
-            { @JoinColumn(name = "tier_ID", nullable = false, updatable = false) },
-              inverseJoinColumns = { @JoinColumn(name = "productRelease_ID", nullable = false, updatable = false) })
+    @JoinTable(name = "tier_has_productReleases", joinColumns = { @JoinColumn(name = "tier_ID", nullable = false, updatable = false) }, inverseJoinColumns = { @JoinColumn(name = "productRelease_ID", nullable = false, updatable = false) })
```
Comment (flopezag) [inline model/src/main/java/com/telefonica/euro_iaas/paasmanager/model/Tier.java]: why do you change this line, you had any problem with the previous one. This new one break the coding style.

File: model/src/main/java/com/telefonica/euro_iaas/paasmanager/model/Tier.java [hunk 23]
```diff
@@ -82,16 +84,11 @@
...
-              inverseJoinColumns = { @JoinColumn(name = "productRelease_ID", nullable = false, updatable = false) })
+    @JoinTable(name = "tier_has_productReleases", joinColumns = { @JoinColumn(name = "tier_ID", nullable = false, updatable = false) }, inverseJoinColumns = { @JoinColumn(name = "productRelease_ID", nullable = false, updatable = false) })
     private List<ProductRelease> productReleases;
 
     @ManyToMany(fetch = FetchType.LAZY)
-    @JoinTable(name = "tier_has_networks", joinColumns =
-            { @JoinColumn(name = "tier_ID", nullable = false, updatable = false) },
-              inverseJoinColumns = { @JoinColumn(name = "network_ID", nullable = false, updatable = false) })
-
+    @JoinTable(name = "tier_has_networks", joinColumns = { @JoinColumn(name = "tier_ID", nullable = false, updatable = false) }, inverseJoinColumns = { @JoinColumn(name = "network_ID", nullable = false, updatable = false) })
```
Comment (flopezag) [inline model/src/main/java/com/telefonica/euro_iaas/paasmanager/model/Tier.java]: idem

File: model/src/main/java/com/telefonica/euro_iaas/paasmanager/model/Tier.java [hunk 24]
```diff
@@ -625,13 +634,18 @@ public String toString() {
         sb.append("[minimumNumberInstances = ").append(this.minimumNumberInstances).append("]");
         sb.append("[initialNumberInstances = ").append(this.initialNumberInstances).append("]");
         sb.append("[payload = ").append(this.payload).append("]");
-        sb.append("[productReleases = ").append(this.productReleases).append("]");
-        sb.append("[networks = ").append(this.networks).append("]");
-        sb.append("[securityGroup = ").append(this.securityGroup).append("]");
+        if (Hibernate.isInitialized(this.productReleases)) {
```
Comment (flopezag) [inline model/src/main/java/com/telefonica/euro_iaas/paasmanager/model/Tier.java]: What do you do if those variables are not initialized in Hibernate?

File: model/src/main/java/com/telefonica/euro_iaas/paasmanager/model/Tier.java (hunk 22 above)
Comment (jesuspg) [inline model/src/main/java/com/telefonica/euro_iaas/paasmanager/model/Tier.java]: a problem with my IDE

**Comment by jframos** [comment]
LGTM

**Pull request merged**